Sources/Positionless/Algorithms/Partition.swift

```diff
@@ -0,0 +1,132 @@
+extension MutableCollection {
+
+  /// Reorders the collection in place so that all elements matching the predicate
+  /// are moved into the suffix, preserving the relative order of only non-matching
+  /// elements. The resulting partition is then projected and passed to `f` with
+  /// returning its result.
+  ///
+  /// - Postcondition: For resulting partition `p`, `p.partitionCount == 2`.
+  ///
+  /// - Complexity: Exactly `count` predicate application. No more than `count` swaps.
+  mutating func halfStablePartition<R>(by predicate: (Element) -> Bool, _ f: (MutableParts) -> R)
+    -> R
+  {
+    return withMutableParts(count: 2) { p in
+      p.halfStablePartition(by: predicate)
+      return f(p)
+    }
+  }
+
+  /// Reorders the collection in place so that all elements matching the predicate
+  /// are moved into the suffix, preserving the relative order of each part.
+  /// The resulting partition is then projected and passed to `f` with returning its result.
+  ///
+  /// - Postcondition: For resulting partition `p`, `p.partitionCount == 2`.
+  ///
+  /// - Complexity: Exactly `count` `predicate` application. At most `count.log2(count)` swaps.
+  mutating func stablePartition<R>(by predicate: (Element) -> Bool, _ f: (MutableParts) -> R)
```

Comment:
```suggestion
  /// Reorders elements so those satisfying `belongsInTail` follow those that
  /// do not, preserving the relative order of elements within each group, and passes
  /// those groups to `f`, returning its result.
  ///
  /// - Complexity: Exactly `count` applications of `belongsInTail`. At most `count * log2(count)` swaps.
  mutating func stablePartition<R>(by belongsInTail: (Element) -> Bool, _ f: (MutableParts) -> R)
```

Are you sure about the complexity specification?

Just curious: Could this be a `subscript(stablePartitionedSoSuffixElementsSatisfy belongsInSuffix: ...) -> MutableParts`?  I think you'd need to make the non-mutating variant trap.